File: mcp-server/internal/domains/skills/tools.go
// Package skills provides the MCP tools and resource templates for the Skill
// domain, backed by the SkillQueryController and SkillCommandController RPCs.
//
// Skills support versioning: the caller can optionally request a specific
// version by tag name (e.g. "stable") or exact SHA-256 hash. When the
// version is omitted, the latest version is returned.
package skills

import (
	"context"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/stigmer/stigmer/mcp-server/internal/domains"
)

// --- get_skill ---

// GetSkillInput defines the parameters for the "get_skill" tool.
type GetSkillInput struct {
	Org     string `json:"org"              jsonschema:"required,description=Organization slug that owns the skill."`
	Slug    string `json:"slug"             jsonschema:"required,description=Skill slug — unique identifier within the org."`
	Version string `json:"version,omitempty" jsonschema:"description=Version to retrieve: tag name (e.g. stable) or SHA-256 hash. Omit for latest."`
}

// Tool returns the MCP tool definition for registration.
func Tool() *mcp.Tool {
	return &mcp.Tool{
		Name: "get_skill",
		Description: "Get full details of a Stigmer skill by org and slug, optionally at a specific version. " +
			"Omit 'version' for the latest.",
	}
}

// Handler returns the typed tool handler. serverAddress is captured at
// registration time; the API key is read from context at call time.
func Handler(serverAddress string) func(context.Context, *mcp.CallToolRequest, *GetSkillInput) (*mcp.CallToolResult, any, error) {
	return func(ctx context.Context, _ *mcp.CallToolRequest, input *GetSkillInput) (*mcp.CallToolResult, any, error) {
		text, err := Fetch(ctx, serverAddress, input.Org, input.Slug, input.Version)
		if err != nil {
			return nil, nil, err
		}
		return domains.TextResult(text)
	}
}

// --- delete_skill ---

// DeleteSkillInput defines the parameters for the "delete_skill" tool.
type DeleteSkillInput struct {
	Org  string `json:"org"  jsonschema:"required,description=Organization slug that owns the skill (e.g. stigmer)."`
	Slug string `json:"slug" jsonschema:"required,description=Skill slug — the unique identifier within the org (e.g. code-review-best-practices). Deletes the skill and all its versions."`
}

// DeleteTool returns the MCP tool definition for the delete_skill tool.
func DeleteTool() *mcp.Tool {
	return &mcp.Tool{
		Name:        "delete_skill",
		Description: "Delete a Stigmer skill and all its versions by org and slug. To create or update skills, use the 'stigmer skill push' CLI command. Returns the deleted skill.",
	}
}

// DeleteHandler returns the typed tool handler for delete_skill.
func DeleteHandler(serverAddress string) func(context.Context, *mcp.CallToolRequest, *DeleteSkillInput) (*mcp.CallToolResult, any, error) {
	return func(ctx context.Context, _ *mcp.CallToolRequest, input *DeleteSkillInput) (*mcp.CallToolResult, any, error) {
		return domains.CallFetch(Delete, ctx, serverAddress, input.Org, input.Slug)
	}
}

File: mcp-server-ts/src/domains/skills/tools.ts
// MCP tools for the Skill domain.
// Go parity: mcp-server/internal/domains/skills/tools.go.
//
// Skills are read-only over MCP (search/get/delete); creation/update happens via
// `stigmer skill push` (content hashing, diffing, tag management). `version` is
// optional — omit it for the latest version.

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import { resolveToken, type BackendTarget } from "../client";
import { textOrError } from "../toolresult";
import { deleteSkill } from "./delete";
import { fetchSkill } from "./fetch";

/** Register every Skill-domain tool; returns the registered tool names. */
export function registerSkillTools(server: McpServer, target: BackendTarget): string[] {
  server.registerTool(
    "get_skill",
    {
      description:
        "Get full details of a Stigmer skill by org and slug, optionally at a specific version. " +
        "Omit 'version' for the latest.",
      inputSchema: {
        org: z.string().describe("Organization slug that owns the skill."),
        slug: z.string().describe("Skill slug — unique identifier within the org."),
        version: z
          .string()
          .optional()
          .describe("Version to retrieve: tag name (e.g. stable) or SHA-256 hash. Omit for latest."),
      },
    },
    (args, extra) =>
      textOrError(() =>
        // Empty string means "latest", matching the Go handler passing input.Version.
        fetchSkill(
          target.serverAddress,
          resolveToken(extra, target.apiKey),
          args.org,
          args.slug,
          args.version ?? "",
        ),
      ),
  );

  server.registerTool(
    "delete_skill",
    {
      description:
        "Delete a Stigmer skill and all its versions by org and slug. To create or update skills, use the 'stigmer skill push' CLI command. Returns the deleted skill.",
      inputSchema: {
        org: z.string().describe("Organization slug that owns the skill (e.g. stigmer)."),
        slug: z
          .string()
          .describe(
            "Skill slug — the unique identifier within the org (e.g. code-review-best-practices). Deletes the skill and all its versions.",
          ),
      },
    },
    (args, extra) =>
      textOrError(() =>
        deleteSkill(target.serverAddress, resolveToken(extra, target.apiKey), args.org, args.slug),
      ),
  );

  return ["get_skill", "delete_skill"];
}
